### PR TITLE
Add a plugin for amdgpu tuning of the `panel_power_savings` attribute

### PR DESCRIPTION
**Pull request by superm1**
This attribute accepts a range from 0 through 4 where larger values will also have larger panel power savings.

Using this has a trade off for color accuracy, and it is only applied when the system is currently operating on battery.

The plugin uses upower to get a signal when battery changes so that it will react instantly.

Intentionally the plugin will check what values are already programmed to the sysfs file to avoid unnecessary writes.  Writing the sysfs file will cause a modeset which isn't necessary if writing the same value twice.

The default values are applied to the profiles that are used in power-profiles-daemon compatbility.  They also match the values used in that software.

### CODE REVIEW

File: tuned/plugins/plugin_amdgpu.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            return None
+
+        # determine value to use (only apply if on battery, otherwise set to 0)
+        if self._on_battery:
+            target = self.target_value
+        else:
+            target = 0
+
+        # make sure the value is different (avoids unnecessary kernel modeset)
+        current = int(self._get_panel_power_savings(device))
```
Comment (zacikpa) [inline tuned/plugins/plugin_amdgpu.py:91]: This fails when the file `/sys/class/drm/card1-eDP-1/amdgpu/panel_power_savings` does not exist.

**Comment by yarda** [comment]
> I'm not sure what I think about enabling this by default for the main stock profiles, mainly due to the new `upower` dependency and the fact that it basically performs a kind of "dynamic tuning", which TuneD has now [abandoned by default](https://github.com/redhat-performance/tuned/pull/603). @yarda, what are your thoughts on this?

Regarding "dynamic tuning" it's not a blocker, because I think we need per plugin configuration, e.g. something like:
```
tuned-main.conf:
dynamic_tuning=no, yes:amdgpu, yes:scheduler
```

This was discussed earlier, but never implemented :). This is for another PR. Regarding the `scheduler` plugin, the dynamic tuning should do the same what currently the `runtime=1` hack do.

**Comment by yarda** [comment]
I think it should reflect the `enable_dbus`  main conf and `--no-dbus` command line option and also work without DBus (e.g. with limited functionality or with plugin disabled), because there are products which runs by default without DBus, e.g. OCP and this shouldn't break it.

**Comment by yarda** [comment]
> I'm not sure what I think about enabling this by default for the main stock profiles, mainly due to the new `upower` dependency

Regarding the upower, it should detect whether upower is available and if yes use it. This is what other plugin do (e.g. disk plugin with hdparm) then we could softdep upower and it will not break consumers without upower.

**Comment by yarda** [comment]
Also what about adding this functionality into the `video` plugin?

**Comment by superm1** [comment]
> > I'm not sure what I think about enabling this by default for the main stock profiles, mainly due to the new `upower` dependency
> 
> Regarding the upower, it should detect whether upower is available and if yes use it. This is what other plugin do (e.g. disk plugin with hdparm) then we could softdep upower and it will not break consumers without upower.

If it's acceptable I would prefer to make power mandatory for the plugin. If it's not available or masked then fail the plugin initialization/load or make it a noop.

I'll double check behavior for the plugin with it masked to make sure I got it right.

**Comment by superm1** [comment]
> Also what about adding this functionality into the `video` plugin?

I was thinking dedicated amdgpu plugin with the idea that we could extend it to other amdgpu stuff later maybe.

**Comment by superm1** [comment]
> I think it should reflect the `enable_dbus`  main conf and `--no-dbus` command line option and also work without DBus (e.g. with limited functionality or with plugin disabled), because there are products which runs by default without DBus, e.g. OCP and this shouldn't break it.

I think plugin should be disabled for this case.

File: tuned/plugins/plugin_amdgpu.py (hunk 1 above)
Comment (superm1) [reply tuned/plugins/plugin_amdgpu.py:91]: But doesn't the plugin fail to add any devices?

Comment (zacikpa) [reply tuned/plugins/plugin_amdgpu.py:91]: Oh, right, that should suffice when implemented correctly. Currently, you do:
```python
self._hardware_inventory.get_devices("drm")
            .match_sys_name("card*eDP*")
            .match_sys_name("amdgpu")
            .match_sys_name("panel_power_savings")
```
Applying `match_sys_name` multiple times yields all devices that match at least one of the names (i.e., it's a union, not an intersection). In my case, `card1-eDP-1` gets matched and added to plugin devices.

Comment (superm1) [reply tuned/plugins/plugin_amdgpu.py:92]: Ah! Got it.  I've pushed a change to only match the eDP and add the device if the sysfs file is found for it.

**Comment by superm1** [comment]
> > I think it should reflect the `enable_dbus`  main conf and `--no-dbus` command line option and also work without DBus (e.g. with limited functionality or with plugin disabled), because there are products which runs by default without DBus, e.g. OCP and this shouldn't break it.
> 
> I think plugin should be disabled for this case.

I believe the modifications I've made will effectively cause this behavior if upower can't be accessed for any reason (such as dbus is disabled).  Let me know if you disagree.

**Review by zacikpa — COMMENTED** [review]
Thanks for the update. Now, this seems to work as expected (i.e., no-op) on machines without AMD GPUs or without upower on the dbus. I'll try to find an AMD GPU machine to test it fully.

**Comment by yarda** [comment]
> > > Also what about adding this functionality into the `video` plugin?
> > 
> > 
> > I was thinking dedicated amdgpu plugin with the idea that we could extend it to other amdgpu stuff later maybe.
> 
> The current functionality of the `video` plugin is also AMD-specific.

I think it could be good extension of the video plugin, because it currently doesn't have dynamic tuning and maybe easier and more user friendly to write more generic:
```
[video]
...
```
than HW specific:
```
[video]
...
[amdgpu]
...
```

**Comment by superm1** [comment]
For now I've force pushed using the old configuration option name and kept the existing style for that plugin.  In general though, I think it would be better to reformat it with `black` if you're open to it.

**Comment by superm1** [comment]
ping?

**Comment by superm1** [comment]
The important part is that it is a mobile design with Vega or newer. I tested it on Phoenix and Rembrandt but I would expect it to be exposed on Renoir or newer.

**Comment by superm1** [comment]
One more thing:

Backport these 2 kernel patches from drm-next to 6.8 kernels to enable access to ABM:

* https://gitlab.freedesktop.org/agd5f/linux/-/commit/63d0b87213a0ba241b3fcfba3fe7b0aed0cd1cc5
* https://gitlab.freedesktop.org/agd5f/linux/-/commit/040fdcde288a2830edc31dd507963d6aadf990d2

They'll be part of 6.9-rc1.

File: tuned/plugins/plugin_video.py [hunk 2]
```diff
@@ -124,3 +199,27 @@ def _get_radeon_powersave(self, device, ignore_missing = False):
...
+		try:
+			value = int(value, 10)
+		except ValueError:
+			if not sim:
+				log.warn("Invalid value %s for panel_power_savings" % value)
+			return None
+		if value in range(0, 5):
+			self.target_value = value
+			if not sim:
+				return self.apply_panel_power_saving_target(device)
```
Comment (zacikpa) [inline tuned/plugins/plugin_video.py:215]: IIUC, this will set up the signaling when handling the first device and subsequently apply the settings for *all* devices because of the loop in `upower_changed`. For the remaining devices, it's then essentially a no-op because they are already configured.

I think that's quite inelegant, but it shouldn't cause issues. Alternatively, you could remove the manual call to `upower_changed` and [not return ](https://github.com/redhat-performance/tuned/pull/601/files#diff-d503f51fb415d704c70263a4c6d4eaf2c2def141d988346452ab610d242a9da5R125)after setting up the signalling (which I guess may require an additional error check).

File: tuned/plugins/plugin_video.py [hunk 3]
```diff
@@ -124,3 +199,27 @@ def _get_radeon_powersave(self, device, ignore_missing = False):
 			return "dpm-" + self._cmd.read_file(sys_files["dpm_state"]).strip()
 		else:
 			return None
+
+	@command_set("panel_power_savings", per_device=True)
+	def _set_panel_power_savings(self, value, device, sim, remove):
```
Comment (zacikpa) [inline tuned/plugins/plugin_video.py:204]: With `sim=True`, the function should do precisely what it would do with `sim=False` except for actually overwriting the file. It can still print warnings and more importantly, it should return the same value. Currently, it always returns `None` if `sim` is `True`.

**Review by zacikpa — COMMENTED** [review]
I still haven't been able to test this, because I can't easily patch the kernel on the AMD GPU machines that I can access.

Nonetheless, I have a couple of comments.

File: tuned/plugins/plugin_video.py (hunk 3 above)
Comment (superm1) [reply tuned/plugins/plugin_video.py]: Fixed

File: tuned/plugins/plugin_video.py (hunk 2 above)
Comment (superm1) [reply tuned/plugins/plugin_video.py]: I did test it on a two GPU system where both GPUs can connect to a mux for eDP and that's why it has the logic it has now; to ensure that you don't have a corner case.  Let me know if you still want it changed.

**Comment by superm1** [comment]
> I still haven't been able to test this, because I can't easily patch the kernel on the AMD GPU machines that I can access.

The patches are going to be in 6.9-rc1 if that makes it easier.

Here is what it looks like when you boot with power plugged in, unplug and then replug:

```
 sudo ./tuned.py  2>&1 | grep amdgpu
2024-03-18 07:53:49,163 INFO     tuned.plugins.base: instance amdgpu: assigning devices card0-eDP-1
2024-03-18 07:53:49,537 INFO     tuned.plugins.plugin_amdgpu: panel_power_savings already 0 [🔋: False]


2024-03-18 07:53:54,324 INFO     tuned.plugins.plugin_amdgpu: panel_power_savings -> 3 [🔋: True]


2024-03-18 07:53:58,661 INFO     tuned.plugins.plugin_amdgpu: panel_power_savings -> 0 [🔋: False]
```

File: tuned/plugins/plugin_video.py (hunk 2 above)
Comment (zacikpa) [reply tuned/plugins/plugin_video.py]: I believe it can stay as it is now.

File: tuned/plugins/plugin_video.py [hunk 4]
```diff
@@ -69,20 +108,53 @@ def _instance_init(self, instance):
...
+		"""Apply the target value to the panel_power_savings file if it doesn't already have it"""
+
+		# if we don't have a proxy, we can't tell if we're on battery
+		if not self.proxy:
+			self.setup_battery_signaling()
+			return None
+
+		# if we don't have the file, we might be radeon not amdgpu
+		if not os.path.exists(self._files(device)["panel_power_savings"]):
+			return None
```
Comment (zacikpa) [inline tuned/plugins/plugin_video.py:129]: I would swap the first two checks to make sure we don't setup the signaling at all if no device supports `panel_power_savings`: 
```suggestion
		# if we don't have the file, we might be radeon not amdgpu
		if not os.path.exists(self._files(device)["panel_power_savings"]):
			return None

		# if we don't have a proxy, we can't tell if we're on battery
		if not self.proxy:
			self.setup_battery_signaling()
			return None
```

**Review by zacikpa — CHANGES_REQUESTED** [review]
Thanks for the update. I've got one more change request and when that is fixed, this looks mergeable to me.

File: tuned/plugins/plugin_video.py [hunk 5]
```diff
@@ -124,3 +199,24 @@ def _get_radeon_powersave(self, device, ignore_missing = False):
...
+			self.target_value = value
+			return self.apply_panel_power_saving_target(device, sim)
+		else:
+			log.warn("Invalid value %s for panel_power_savings" % value)
+		return None
+
+	@command_get("panel_power_savings")
+	def _get_panel_power_savings(self, device, ignore_missing=False):
+		"""Get the current panel_power_savings value"""
+		fname = self._files(device)["panel_power_savings"]
```
Comment (zacikpa) [inline tuned/plugins/plugin_video.py:221]: Check that the file exists the same way as when attempting to set it. This is necessary for the `tuned-adm verify` command to succeed when the device does not support `panel_power_savings`.
```suggestion
		"""Get the current panel_power_savings value"""
		if not os.path.exists(self._files(device)["panel_power_savings"]):
			log.debug("panel_power_savings is not supported on '%s'" % device)
			return None
		fname = self._files(device)["panel_power_savings"]
```

**Review by zacikpa — COMMENTED** [review]
Managed to find one other thing still.

File: tuned/plugins/plugin_video.py (hunk 5 above)
Comment (superm1) [reply tuned/plugins/plugin_video.py]: Thx, looks good to me.

**Review by zacikpa — APPROVED** [review]
Thanks. LGTM, so approving, but leaving the final merge decision on @yarda.

**Comment by superm1** [comment]
@yarda ping?

**Review by zacikpa — CHANGES_REQUESTED** [review]
As discussed in #616, the response to battery/AC changes should be handled by tuned-ppd and thus moved to that PR.

File: tuned/plugins/plugin_video.py [hunk 6]
```diff
@@ -5,15 +5,21 @@
...
 import errno
 import re
+import dbus
 
 log = tuned.logs.get()
 
+UPOWER_DBUS_NAME = "org.freedesktop.UPower"
+UPOWER_DBUS_PATH = "/org/freedesktop/UPower"
+UPOWER_DBUS_INTERFACE = "org.freedesktop.UPower"
+
```
Comment (zacikpa) [inline tuned/plugins/plugin_video.py:15]:
```suggestion
```

File: tuned/plugins/plugin_video.py [hunk 7]
```diff
@@ -5,15 +5,21 @@
 import os
 import errno
 import re
+import dbus
```
Comment (zacikpa) [inline tuned/plugins/plugin_video.py:8]:
```suggestion
```

File: tuned/plugins/plugin_video.py [hunk 8]
```diff
@@ -40,17 +46,28 @@ class VideoPlugin(base.Plugin):
...
+
+	Mobile hardware with amdgpu driven eDP panels can be configured
+	with the [option]`panel_power_savings` option.
+	This accepts a value range from 0 to 4, where 4 is the highest power savings
+	but will trade off color accuracy.
 	"""
 
+	def __init__(self, *args, **kwargs):
+		super(VideoPlugin, self).__init__(*args, **kwargs)
+		self.proxy = None
```
Comment (zacikpa) [inline tuned/plugins/plugin_video.py:58]:
```suggestion
```

**Review by zacikpa — CHANGES_REQUESTED** [review]
Thanks, please delete the remnants of the dbus functionality according to the comments.

**Review by zacikpa — APPROVED** [review]

**Comment by superm1** [comment]
> Thanks, please remove the upower mention from the commit https://github.com/redhat-performance/tuned/commit/9d108694d623fb7d32c7dde40881041e215ff6ce description:

Rebased on main and done now.

**Review by yarda — APPROVED** [review]

**Comment by yarda** [comment]
Thanks.
